File: programs/ch10/prolog/eliza.pl
:- [tokenize_dcg].
% % A simplified version of the ELIZA program
% %

% The main loop reads the input and calls process/1
% It stops when the input is the word bye.

eliza :-
	write('Hello, I am ELIZA. How can I help you?'), nl,
	repeat,
	write('> '),
	tokenize_line(In),
	process(In).


% process/1 accepts the userís utterance. It either terminates
% or tries to find a template matching to the userís
% utterance

process([bye | _]) :-
	write('ELIZA: bye'), nl, !.
process(In) :-
	utterance(Out, In, []),
	!,
	write('ELIZA: '), write_answer(Out),
	fail.

% utterance is a DCG rule describing a set of templates

utterance(['Why', aren, '''', t, you | Y]) -->
	['I', am, not], end(Y).
utterance(['How', long, have, you, been | Y]) -->
	['I', am], end(Y).
utterance(['Why', do, you, like | Y]) -->
	['I', like], end(Y).
utterance(['Do', you, often, think, of | Y]) -->
	beginning(_), ['I', remember], end(Y).
utterance(['Please', tell, me, more, about, your, X, .]) -->
	beginning(_), [X], end(_), {important(X)}.
utterance(['Why', are, so, negative, '?']) -->
	[no].
utterance(['Tell', me, more, .]) -->
	[_].
utterance(['Please', go, on, .]) -->
	beginning(_).

% The Prolog clauses beginning/3 qnd end/3
% to link the pieces of the utterance

beginning(X, Y, Z) :- append(X, Z, Y).

end(X, Y, Z) :- append(X, Z, Y).

% The Freudian facts
important(father).
important(mother).
important(brother).
important(sister).

% write_answer/1 prints the content of the list containing
% Eliza's answer.

write_answer([Head | Tail]) :-
	write(Head), write(' '),
	write_answer(Tail).
write_answer([]) :- nl.

File: programs/ch10/python/eliza.py
"""
A simplified implementation of ELIZA
"""
__author__ = "Pierre Nugues"

import regex as re
import numpy as np

dialogue_pairs = {
    '.*I am not (.+)': [
        r'Why aren\'t you \1'
    ],
    '.*I am (.+)': [
        r'How long have you been \1'
    ],
    '.*I like (.+)': [
        r'Why do you like \1'
    ],
    '.*I remember (.+)': [
        r'Do you often think of \1',
        'What else do you remember',
        r'Why do you remember \1 just now'
    ],
    '.*(father|mother|brother|sister).*': [
        r'Please tell me more about your \1'
    ],
    '^no$': [
        'Why are you so negative?',
        'Why not?'
    ],
    '^\p{L}+$': [
        'Tell me more...'
    ],
    # When nothing matches
    '.*': [
        'I am not sure I understand you fully',
        'Please go on',
        'What does that suggest to you'
    ]
}


def match_utterance(user_input):
    """
    match_utterance accepts the user's utterance and
    tries to find a template matching it
    """
    for key in dialogue_pairs:
        matches = list(re.finditer(key, user_input))
        if matches:
            choice = np.random.randint(len(dialogue_pairs[key]))
            answer = re.sub(key, dialogue_pairs[key][choice],
                            user_input, count=1)
            return answer


# The main loop reads the input and calls match_utterance
# It stops when the input is the word bye.
print('Hello, I am ELIZA. How can I help you?')
user_input = input('User: ')
while user_input.lower() != 'bye':
    answer = match_utterance(user_input)
    print('Eliza:', answer)
    user_input = input('User: ')
print('Eliza:', 'Bye')
